### PR TITLE
Initial support for arrays

### PR DESCRIPTION
**Pull request by akshayjshah**
Add basic support for logging arrays. To limit the scope of this commit,
it introduces only arrays of ObjectMarshalers, arrays of arrays, and
arrays of bools. It should, however, clarify how I intend to support
arrays of other types.

Of note, I opted not to have `zap.Bools` return a `Field`. Instead, it
returns a `zapcore.ArrayMarshaler`. This lets us build arrays of
heterogenous arrays. For most users, this will be papered over by the
upcoming `zap.Any(key string, val interface{}) zapcore.Field` function,
which will basically be a huge type switch.

(Along the way, I broke out our multi-error implementation into a separate
internal package.)

This starts to address #136, and is an alternative to the approach in #211.

From the small benchmarks included in this PR:
```
BenchmarkBoolsArrayMarshaler-4   	 1000000	      1028 ns/op	    1088 B/op	       3 allocs/op
BenchmarkBoolsReflect-4          	  500000	      3293 ns/op	    2280 B/op	       8 allocs/op
```
All allocations in the ArrayMarshaler benchmark come from allocating byte slices for encoders.

### CODE REVIEW

File: zapcore/encoder.go [hunk 1]
```diff
@@ -22,6 +22,31 @@ package zapcore
...
+	AddFloat64(key string, value float64)
+	AddInt64(key string, value int64)
+	AddUint64(key string, value uint64)
+	AddObject(key string, marshaler ObjectMarshaler) error
+	AddArray(key string, marshaler ArrayMarshaler) error
+	// AddReflected uses reflection to serialize arbitrary objects, so it's slow
+	// and allocation-heavy.
+	AddReflected(key string, value interface{}) error
+	AddString(key, value string)
+}
```
Comment (akshayjshah) [inline zapcore/encoder.go]: Moved most of this  here from `object_encoder.go`, since it seemed silly to have two interface-only files for encoders.

File: array.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+	"go.uber.org/zap/internal/multierror"
+	"go.uber.org/zap/zapcore"
+)
+
+// TODO: actually add an example showing off how to use these wrappers with the
+// Array field constructor.
+
+// Arrays wraps a slice of ArrayMarshalers so that it satisfies the
+// ArrayMarshaler interface. See the Array function for a usage example.
+func Arrays(as []zapcore.ArrayMarshaler) zapcore.ArrayMarshaler {
```
Comment (prashantv) [inline array.go]: I don't think `Arrays` and `Objects` is going to be that useful to users.

If someone is implementing the `ArrayMarshaler` interface, then they probably care about performance, and wouldn't want to build up a separate new slice of `[]zapcore.ArrayMarshaler`.

What use case are you expecting for `Arrays` and `Objects`?

File: array_test.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package zap
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
```
Comment (prashantv) [inline array_test.go]: nit: import ordering

File: array.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+func Bools(bs []bool) zapcore.ArrayMarshaler {
+	return bools(bs)
+}
+
+type arrayMarshalers []zapcore.ArrayMarshaler
+
+func (as arrayMarshalers) MarshalLogArray(arr zapcore.ArrayEncoder) error {
+	var errs *multierror.Error
+	for i := range as {
+		if as[i] == nil {
```
Comment (prashantv) [inline array.go]: If we want to keep `arrayMarshalers` around, we should consider whether ignoring `nil` is the right behaviour. Should it just be an error? I think I'm learning towards remove `Arrays` and `Objects` and keep things minimal. It's easy enough for others to implement and add later.

If we really want to keep this type, can we add unit tests for the `== nil` branch here and in `Objects`?

File: internal/multierror/multierror.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// Package multierror provides a simple way to treat a collection of errors as
+// a single error.
+package multierror
+
+import "go.uber.org/zap/buffers"
```
Comment (prashantv) [inline internal/multierror/multierror.go]: is this package not internal on purpose? I really think this should be internal. Otherwise one bad user of `buffers` will cause all sorts of weird data races / corrupt log messages. I really don't think we want any other library to be able to put buffers into our pool.

File: internal/multierror/multierror.go [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+	}
+}
+
+// Append adds an error to the collection. Adding a nil error is a no-op.
+func (e *Error) Append(err error) *Error {
+	if err == nil {
+		return e
+	}
+	if e == nil {
+		e = &Error{}
```
Comment (prashantv) [inline internal/multierror/multierror.go]: This is a strange pattern, you have an `Append` that works on a pointer, sometimes modifies the value inside, sometimes returns a new pointer. Given that, why do you need this method on the pointer? Why not always work on the value? Passing around a `[]error` by value is not expensive, makes the semantics cleaner, and reduces the need for `*Error`

That'll probably also avoid an allocation when there is an error

File: internal/multierror/multierror_test.go [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+package multierror
+
+import (
+	"errors"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
```
Comment (prashantv) [inline internal/multierror/multierror_test.go]: One slightly odd test that I think is worth having -- you really don't want `Error` to implement the `error` interface, since that could cause odd issues. So we could have a test like:
```go
var v interface{} = Error{}
_, ok := v.(error)
require.False(t, ok)
```

File: zapcore/json_encoder_test.go [hunk 8]
```diff
@@ -136,6 +199,57 @@ func TestJSONEncoderFields(t *testing.T) {
 	}
 }
 
+func TestJSONEncoderArrayTypes(t *testing.T) {
```
Comment (prashantv) [inline zapcore/json_encoder_test.go]: Can you add unit tests for:
 - empty array
 - array with different types

File: array_test.go [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+			}),
+			[]interface{}{
+				[]interface{}{true},
+				[]interface{}{false},
+			},
+		},
+		{
+			"objects",
+			Objects([]zapcore.ObjectMarshaler{
+				zapcore.ObjectMarshalerFunc(func(_ zapcore.ObjectEncoder) error { return nil }),
```
Comment (prashantv) [inline array_test.go]: here and below, since you don't use any args, you can just make this `func(zapcore.ObjectEncoder) error { .. }`

File: array.go (hunk 4 above)
Comment (akshayjshah) [inline array.go]: Sure, I can add some tests. I'm going to have to come through lots and lots of `zapcore` and improve test coverage over the next week.

File: array.go (hunk 2 above)
Comment (akshayjshah) [inline array.go]: `Arrays` is, IMO, useful precisely because most users won't be implementing `ArrayMarshaler` themselves - they'll be using the umpteen wrapper types we're providing, like `Bools`. Since we're already going down a road where we support mixed-type arrays, I'd like to take that to its logical conclusion.

I'll email you about `Objects` - relies on some internal examples.

File: internal/multierror/multierror.go (hunk 5 above)
Comment (akshayjshah) [inline internal/multierror/multierror.go]: Yep, makes sense. My bad.

File: zapcore/json_encoder_bench_test.go [hunk 10]
```diff
@@ -63,8 +63,8 @@ func BenchmarkZapJSON(b *testing.B) {
 		for pb.Next() {
 			enc := newJSONEncoder(cfg)
 			enc.AddString("str", "foo")
-			enc.AddInt64("int", 1)
-			enc.AddInt64("int64", 1)
+			enc.AddInt64("int64-1", 1)
```
Comment (jcorbin) [inline zapcore/json_encoder_bench_test.go]: What's up with this, it seems unrelated?

File: zapcore/memory_encoder_test.go [hunk 11]
```diff
@@ -27,22 +27,45 @@ import (
...
+			e.AppendBool(true)
+			return nil
+		}))
+		return nil
+	})), "Expected AddArray to succeed.")
+	// Nested objects and arrays.
+	assert.NoError(t, enc.AddArray("turduckens", turduckens(2)), "Expected AddObject to succeed.")
+	assert.NoError(t, enc.AddObject("turducken", turducken{}), "Expected AddObject to succeed.")
 
+	wantTurducken := MapObjectEncoder{
```
Comment (jcorbin) [inline zapcore/memory_encoder_test.go]: +100

File: internal/multierror/multierror.go [hunk 12]
```diff
@@ -0,0 +1,71 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
```
Comment (akshayjshah) [inline internal/multierror/multierror.go]: My understanding is that copyright applies to the creative work as a whole, which is the zap project; since that project started in 2016, all copyright starts there. (Imagine that Walt Disney updates a few of the line arcs in the original Mickey Mouse drawing. Those lines don't have a separate copyright date.)

Of course, I could easily be wrong.

File: internal/multierror/multierror.go [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+// Error wraps a []error to implement the error interface.
+type Error struct {
+	errs errSlice
+}
+
+// AsError converts the collection to a single error value.
+//
+// Note that failing to use AsError will almost certainly lead to bugs with
+// non-nil interfaces containing nil concrete values.
+func (e Error) AsError() error {
```
Comment (akshayjshah) [inline internal/multierror/multierror.go]: That is indeed the use case. While I agree that error handling is often repetitive, I think we're often encouraged to program with error values. Where possible, I'd like to make things less repetitive. (I'm treating `bufio.Reader` as my spirit animal here.)

File: zapcore/json_encoder_bench_test.go (hunk 10 above)
Comment (akshayjshah) [inline zapcore/json_encoder_bench_test.go]: Yep, the key was just incorrect - it's not an int, but an int64.

File: zapcore/memory_encoder_test.go (hunk 11 above)
Comment (akshayjshah) [inline zapcore/memory_encoder_test.go]: 🦃

File: zapcore/json_encoder_bench_test.go (hunk 10 above)
Comment (akshayjshah) [inline zapcore/json_encoder_bench_test.go]: Yep, the key was just wrong. Noticed and fixed in passing.
